sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/SetFns.java

```diff
@@ -187,83 +237,112 @@
-
-    PCollection<KV<T, CoGbkResult>> coGbkResults =
-        KeyedPCollectionTuple.of(leftCollectionTag, left)
-            .and(rightCollectionTag, right)
-            .apply(CoGroupByKey.create());
-    // TODO: lift combiners through the CoGBK.
-    return coGbkResults.apply(
-        ParDo.of(
-            new DoFn<KV<T, CoGbkResult>, T>() {
-
-              @ProcessElement
-              public void processElement(ProcessContext c) {
-                KV<T, CoGbkResult> elementGroups = c.element();
-
-                CoGbkResult value = elementGroups.getValue();
-                long inFirstSize = Iterables.size(value.getAll(leftCollectionTag));
-                long inSecondSize = Iterables.size(value.getAll(rightCollectionTag));
-
-                T element = elementGroups.getKey();
-                for (long i = 0L; i < fn.apply(inFirstSize, inSecondSize); i++) {
-                  c.output(element);
-                }
-              }
-            }));
+  public static <T> Flatten.PCollections<T> unionAll() {
+    return Flatten.pCollections();
   }
 
   public static class SetImpl<T> extends PTransform<PCollection<T>, PCollection<T>> {
-    private final PCollection<T> rightCollection;
-    private final SerializableBiFunction<Long, Long, Long> fn;
 
-    private SetImpl(PCollection<T> rightCollection, SerializableBiFunction<Long, Long, Long> fn) {
-      this.rightCollection = rightCollection;
-      this.fn = fn;
+    private final transient PCollection<T> right;
+    private final PTransform<PCollectionList<T>, PCollection<T>> listTransformFn;
+
+    private SetImpl(
+        PCollection<T> rightCollection,
+        PTransform<PCollectionList<T>, PCollection<T>> listTransformFn) {
+      this.right = rightCollection;
+      this.listTransformFn = listTransformFn;
     }
 
     @Override
     public PCollection<T> expand(PCollection<T> leftCollection) {
-      return performSetOperation(leftCollection, rightCollection, fn)
-          .setCoder(leftCollection.getCoder());
+      return PCollectionList.of(leftCollection).and(right).apply(listTransformFn);
     }
   }
 
-  public static class SetUnionAllImpl<T> extends PTransform<PCollection<T>, PCollection<T>> {
-    private final PCollection<T> rightCollection;
+  public static class SetImplCollections<T> extends PTransform<PCollectionList<T>, PCollection<T>> {
+
+    private final transient SerializableBiFunction<Long, Long, Long> fn;
+
+    private SetImplCollections(SerializableBiFunction<Long, Long, Long> fn) {
+      this.fn = fn;
+    }
+
+    private static <T> PCollection<T> performSetOperationCollectionList(
+            PCollectionList<T> inputs, SerializableBiFunction<Long, Long, Long> fn) {
+      List<PCollection<T>> all = inputs.getAll();
+      int size = all.size();
+      if (size == 1) {
+        return inputs.get(0); // Handle only one PCollection in list. Coder is already specified
```

Comment:
This is only true for the all variants, right?

Reply:
Yes. I need to check this as i can't bindly call `setCoder` if there is only one element in the list.

Reply:
I have removed this check now.